File: 999.go
func numRookCaptures(board [][]byte) int {
	var rookI, rookJ int
outer:
	for i := range board {
		for j := range board[i] {
			if board[i][j] == 'R' {
				rookI, rookJ = i, j
				break outer
			}
		}
	}

	var result int
	i := rookI - 1
	for ; i >= 0 && board[i][rookJ] == '.'; i-- {
	}
	if i >= 0 && board[i][rookJ] == 'p' {
		result++
	}

	i = rookI + 1
	for ; i < len(board) && board[i][rookJ] == '.'; i++ {
	}
	if i < len(board) && board[i][rookJ] == 'p' {
		result++
	}

	j := rookJ - 1
	for ; j >= 0 && board[rookI][j] == '.'; j-- {
	}
	if j >= 0 && board[rookI][j] == 'p' {
		result++
	}

	j = rookJ + 1
	for ; j < len(board[rookI]) && board[rookI][j] == '.'; j++ {
	}
	if j < len(board[rookI]) && board[rookI][j] == 'p' {
		result++
	}

	return result
}

File: 999.rs
impl Solution {
    pub fn num_rook_captures(board: Vec<Vec<char>>) -> i32 {
        let (rook_i, rook_j) = Self::find_rook_pos(&board);

        let mut result = 0;
        let mut i = rook_i - 1;
        while i < board.len() && board[i][rook_j] == '.' {
            i -= 1;
        }
        if i < board.len() && board[i][rook_j] == 'p' {
            result += 1;
        }

        i = rook_i + 1;
        while i < board.len() && board[i][rook_j] == '.' {
            i += 1;
        }
        if i < board.len() && board[i][rook_j] == 'p' {
            result += 1;
        }

        let mut j = rook_j - 1;
        while j < board[rook_i].len() && board[rook_i][j] == '.' {
            j -= 1;
        }
        if j < board[rook_i].len() && board[rook_i][j] == 'p' {
            result += 1;
        }

        j = rook_j + 1;
        while j < board[rook_i].len() && board[rook_i][j] == '.' {
            j += 1;
        }
        if j < board[rook_i].len() && board[rook_i][j] == 'p' {
            result += 1;
        }

        result
    }

    fn find_rook_pos(board: &Vec<Vec<char>>) -> (usize, usize) {
        let mut rook_i = 0;
        let mut rook_j = 0;
        
        for (i, row) in board.iter().enumerate() {
            for (j, c) in row.iter().enumerate() {
                if *c == 'R' {
                    return (i, j);
                }
            }
        }

        unreachable!()
    }
}
